dds/DCPS/XTypes/TypeObject.h

```diff
@@ -64,32 +64,83 @@ namespace XTypes {
   const DCPS::Encoding& get_typeobject_encoding();
 
   template <typename T>
-  struct Optional {
-    bool present;
-    T value;
-
+  class Optional {
+  public:
     Optional()
-      : present(false)
+      : present_(false),
+        value_()
     {}
 
     Optional(const T& v)
-      : present(true)
-      , value(v)
-    {}
+      : present_(true)
+    {
+      new(value_) T(v);
+    }
+
+    ~Optional() {
+      if (present_) {
+        value().~T();
+      }
+    }
+
+    Optional(const Optional& rhs)
+      : present_(false) {
+      *this = rhs;
+    }
+
+    Optional& operator=(const Optional& rhs) {
+      if (this != &rhs) {
+        if (present_) {
+          if (rhs.present_) {
+            value() = rhs.value();
+          }
+          else {
+            value().~T();
+          }
+        }
+        else {
+          if (rhs.present_) {
+            new(value_) T(rhs.value());
+          }
+        }
+        present_ = rhs.present_;
+      }
+      return *this;
+    }
 
     bool operator==(const Optional& other) const
     {
-      if (present) {
-        return present == other.present && value == other.value;
+      if (present_) {
+        return present_ == other.present_ && value() == other.value();
       }
 
-      return present == other.present;
+      return present_ == other.present_;
     }
 
     bool operator!=(const Optional& other) const
     {
       return !(*this == other);
     }
+
+    operator bool() const {
+      return present_;
+    }
+
+    bool has_value() const {
+      return present_;
+    }
+
+    T& value() {
+      return reinterpret_cast<T&>(value_);
+    }
+
+    const T& value() const {
+      return reinterpret_cast<const T&>(value_);
+    }
+
+  private:
+    bool present_;
+    unsigned char value_[sizeof(T)];
```

Comment:
Like the one in AnnotationParameterValue (in this file), I think this needs to handle alignment:
```
    union {
      ACE_CDR::LongDouble max_alignment;
      unsigned char value_[sizeof(T)];
    };
```

Reply:
We have a similar union member defined as a `ACE_CDR::ULongLong` in `TypeIdentifier`. I'm wondering what is the purpose of this member and why some other anonymous union, like `AnnotationParameterValue`, doesn't have it?

Reply:
The types used in TypeIdentifier have different alignment requirements.
AnnotationParameterValue does have it in the current version.

Reply:
Yeah, sorry, it's in `AnnotationParameterValue`. My local branch is not up to date.